### PR TITLE
caculate the file metadata changes

### PR DESCRIPTION
**Pull request by partychen**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/FileMetadataHelper.cs [hunk 1]
```diff
@@ -3,16 +3,133 @@
...
-            // TODO: implement this
-            yield break;
+            if (left == null && right == null)
+            {
+                return new List<GlobMatcher>();
+            }
+            if (left == null || right == null)
+            {
+                return left == null
+                    ? right.SelectMany(r => r.Value).Select(r => r.Glob)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/FileMetadataHelper.cs]: What about adding a method `FileMetadata.GetAllGlobs()`?

Comment (partychen) [reply src/Microsoft.DocAsCode.Build.Engine/FileMetadataHelper.cs]: good suggestion.

File: src/Microsoft.DocAsCode.Build.Engine/FileMetadataHelper.cs [hunk 2]
```diff
@@ -3,16 +3,133 @@
...
+                if (!commonItems.Any(i => i.Equals(rightItem)))
+                {
+                    changes.Add(rightItem.Glob);
+                }
+            }
+
+            return changes;
+        }
+
+        public static IList<T> GetLongestCommonSequence<T>(ImmutableArray<T> leftItems, ImmutableArray<T> rightItems)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/FileMetadataHelper.cs]: Looks like this method is generic enough to move to `CollectionUtility`

File: test/Microsoft.DocAsCode.Build.Engine.Tests/FileMetadataHelperTest.cs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                    new FileMetadataItem(new GlobMatcher("/[]\\*.cs"), "meta", new Dictionary<string, object> { ["key"] = "2" }),
+                    new FileMetadataItem(new GlobMatcher("*/*.cs"), "meta", new object[] { "1", "2" }),
+                    new FileMetadataItem(new GlobMatcher("**"), "meta", new Dictionary<string, object> { ["key"] = new object[] { "1", "2" } })
+                )
+            });
+
+
+            var actual = FileMetadataHelper.GetChangedGlobs(left, right);
+            Assert.True(actual != null);
+            Assert.True(actual.Count() == 0);
```
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/FileMetadataHelperTest.cs]: minor: can use `Assert.NotNull` and `Assert.Empty`

File: test/Microsoft.DocAsCode.Build.Engine.Tests/FileMetadataHelperTest.cs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                    new FileMetadataItem(new GlobMatcher("*.m"), "meta", true),
+                    new FileMetadataItem(new GlobMatcher("abc"), "meta", "string"),
+                    new FileMetadataItem(new GlobMatcher("/[]\\*.cs"), "meta", new Dictionary<string, object> { ["key"] = "2" }),
+                    new FileMetadataItem(new GlobMatcher("*/*.cs"), "meta", new object[] { "1", "2" }),
+                    new FileMetadataItem(new GlobMatcher("**"), "meta", new Dictionary<string, object> { ["key"] = new object[] { "1", "2" } })
+                )
+            });
+
+
+            var actual = FileMetadataHelper.GetChangedGlobs(left, right);
```
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/FileMetadataHelperTest.cs]: `ToList()`

File: test/Microsoft.DocAsCode.Build.Engine.Tests/IncrementalBuildTest.cs [hunk 5]
```diff
@@ -5077,7 +5077,7 @@ public void TestIncrementalWithFileMetadataChange()
             });
 
             FileCollection files = new FileCollection(Directory.GetCurrentDirectory());
-            files.Add(DocumentType.Article, new[] { fileAddFm, fileRemoveFm, fileModifyFm, fileKeepFm});
+            files.Add(DocumentType.Article, new[] { fileAddFm, fileRemoveFm, fileModifyFm, fileKeepFm });
```
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/IncrementalBuildTest.cs]: What about adding a `fileReorderFm` to test order case?

**Pull request merged**